src/hypothesis/strategies.py

```diff
@@ -1155,23 +1155,36 @@ def decimals(min_value=None, max_value=None,
     # Set up a strategy for finite decimals
     if places is not None:
         # Fixed-point decimals are basically integers with a scale factor
+        # and annoying tendencies to break in unusual precision contexts.
+        def ctx(val):
+            """Return a context in which this value is lossless."""
+            precision = ceil(math.log10(abs(val or 1))) + places + 1
+            return Context(prec=max([precision, 1]))
+
+        def int_to_decimal(val):
+            context = ctx(val)
+            return context.quantize(context.multiply(val, factor), factor)
 
-        def try_quantize(d):
-            try:
-                return d.quantize(factor)
-            except InvalidOperation:  # pragma: no cover
-                return None
         factor = Decimal(10) ** -places
-        max_num = max_value / factor if max_value is not None else None
-        min_num = min_value / factor if min_value is not None else None
-        strat = integers(min_value=min_num, max_value=max_num)\
-            .map(lambda d: try_quantize(d * factor))\
-            .filter(lambda d: d is not None)
+        min_num, max_num = None, None
+        if min_value is not None:
+            min_num = ceil(ctx(min_value).divide(min_value, factor))
+        if max_value is not None:
+            max_num = floor(ctx(max_value).divide(max_value, factor))
+        if None not in (min_num, max_num) and min_num > max_num:
+            raise InvalidArgument(
+                'There are no decimals between min_value=%s and max_value=%s '
```

Comment:
ultra ultra nitpicking feel free to ignore: I think this reads slightly better as `'There are no decimals with at most %d places between min_value=%s and max_value=%s'` (i.e. swapping the places and the bounds)